ui/src/message_popup/popup_question.jsx

```diff
@@ -53,25 +56,68 @@ export default React.createClass({
   },
 
   onTextChanged({target:{value}}:TextChangeEvent) {
-    this.setState({ responseText: value });
+    this.setState({ initialResponseText: value });
   },
 
-  onSendPressed() {
-    const {elapsedMs, responseText} = this.state;
-    const {question, shouldShowStudentCard, allowedToToggleHint} = this.props;
+  onSavePressed() {
+    const {helpType} = this.props;
+    this.logResponse();
+    if(helpType === 'feedback'){
+      this.setState({isRevising:true});
+    }else{
+      this.props.onDone();
+    }
+  },
+  
+  onDonePressed(revised, text) {
+    if(revised){
+      this.logRevision(text);
+    }else{
+      this.logRevisionDeclined();
+    }
+    this.props.onDone();
+  },
+  
+  logResponse() {
+    this.logData(undefined);
+  },
+  
+  logRevision(finalText) {
+    this.logData(finalText);
+  },
+  
+  logRevisionDeclined(){
+    this.logData(undefined);
+  },
+  
+  logData(value) {
```

Comment:
I'd suggest simplifying the branching here a bit more, and having `logData` take the type of message to record, and then optional params that are included in the response (eg., `finalText` for `logRevision`).